### PR TITLE
Closed campaigns logic

### LINKED ISSUE #308: Think about closed campaign logic

### LINKED ISSUE #381: Fix expiry date / end dates being start of day
Changed to use Date.current. We should think about changing the end/expiry dates to be 23:59 so that the expiry date makes sense (currently a coupon that expires at 2 November 00:00 is displayed as 2 Nov, but normal people will perceive that to be till 2 November 23:59). Will create an issue for that.

_Originally posted by @ganhongyao in https://github.com/Giving-Coupons/giving-coupons/pull/373#discussion_r1011750603_

### PR DESCRIPTION
**Pull request by ganhongyao**
### Campaign ended (redemption)
image

### Campaign ended (contribution)
image

### Campaign not started (both redemption and contribution)
image

Closes #308
Closes #381

### CODE REVIEW

**Review by zognin — COMMENTED** [review]
Is this meant to be allowed? Right now this works
Screenshot 2022-11-04 at 8 29 31 PM

May want to sort non red entries on top/ use 2 tabs. I didn't know what the red meant at first
Screenshot 2022-11-04 at 8 38 28 PM

File: backend/app/models/campaign.rb [hunk 1]
```diff
@@ -99,4 +100,12 @@ def approve_associated_interest
 
     interest.approve
   end
+
+  def update_coupons_expiry_if_needed
```
[the remark this answers is not in the document]
Comment (ganhongyao) [reply backend/app/models/campaign.rb:104]: Yes, that's meant to happen as we don't want a state where the coupon expires later than the campaign.

File: backend/app/models/redemption.rb [hunk 2]
```diff
@@ -39,4 +40,15 @@ def coupon_not_expired
...
     errors.add(:coupon, 'has already expired')
   end
+
+  def redemption_when_campaign_active
+    return if campaign_charity.nil?
+
+    campaign = campaign_charity.campaign
+    return if campaign.nil?
+
+    return if campaign.start <= redeemed_at && redeemed_at <= campaign.end
```
[the remark this answers is not in the document]
Comment (ganhongyao) [reply backend/app/models/redemption.rb:50]: Yes, this accounts for time. Also currently our table stores campaign.start as start of day and campaign.end as end of day.

**Comment by zognin** [comment]
Oh yea are you going to remove the contribute buttons on the campaign list page? for upcoming and completed campaigns

File: frontend/components/redeem/CampaignEndedDisplay.tsx [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+
+      <Button
+        actionType="secondary"
+        onClick={() => {
+          log("[CampaignEndedDisplay] Click 'View Others'");
+          router.push(`/campaigns`);
+        }}
+      >
+        Browse Active Campaigns
+      </Button>
```
[the remark this answers is not in the document]
Comment (ganhongyao) [reply frontend/components/redeem/CampaignEndedDisplay.tsx:61]: image

image

Fixed

**Comment by ganhongyao** [comment]
image

@zognin regarding your comment on the red rows, I separated the coupons into 3 separate tabs
- Unredeemed (unexpired and unredeemed)
- Redeemed
- Expired (expired and unredeemed)

**Review by zognin — APPROVED** [review]
LGTM can merge when conflicts are resolved

**Pull request merged**